### ISSUE TITLE
Dynamic reloading of TLS client certificates

### ISSUE THREAD

**Issue by cbroglie** [cloudprober/cloudprober#405]
**Describe the feature you'd like and the problem it will solve**

cloudprober reads TLS certs once at init, and then uses them forever. Our environments are moving towards short-lived certs (e.g. 24h). These certs are renewed automatically by [cert-manager](https://www.cncf.io/projects/cert-manager/) and updated on disk by k8s, but cloudprober doesn't know to reload them. We can work around the problem by triggering restarts of cloudprober, but it would be nice if cloudprober supported dynamic reloading directly.

One way to add this support is via [certinel](https://github.com/cloudflare/certinel). The patch for HTTP probes would look something like:
```diff
diff --git a/probes/http/http.go b/probes/http/http.go
index 5b39613..13331e3 100644
--- a/probes/http/http.go
+++ b/probes/http/http.go
@@ -21,6 +21,7 @@ import (
 	"crypto/tls"
 	"fmt"
 	"io/ioutil"
+	"log"
 	"math/rand"
 	"net"
 	"net/http"
@@ -32,6 +33,7 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/cloudflare/certinel/fswatcher"
 	"github.com/cloudprober/cloudprober/common/httputils"
 	"github.com/cloudprober/cloudprober/common/oauth"
 	"github.com/cloudprober/cloudprober/common/tlsconfig"
@@ -62,6 +64,8 @@ type Probe struct {
 	c    *configpb.ProbeConf
 	l    *logger.Logger
 
+	certinel *fswatcher.Sentinel
+
 	baseTransport http.RoundTripper
 	redirectFunc  func(req *http.Request, via []*http.Request) error
 
@@ -135,6 +139,10 @@ func (p *Probe) getTransport() (*http.Transport, error) {
 				return nil, err
 			}
 		}
+
+		if p.certinel != nil {
+			transport.TLSClientConfig.GetClientCertificate = p.certinel.GetClientCertificate
+		}
 	}
 
 	// If HTTP keep-alives are not enabled (default), disable HTTP keep-alive in
@@ -201,6 +209,14 @@ func (p *Probe) Init(name string, opts *options.Options) error {
 		p.oauthTS = oauthTS
 	}
 
+	if c.TlsConfig.GetTlsCertFile() != "" {
+		certinel, err := fswatcher.New(c.TlsConfig.GetTlsCertFile(), c.TlsConfig.GetTlsKeyFile())
+		if err != nil {
+			return err
+		}
+		p.certinel = certinel
+	}
+
 	transport, err := p.getTransport()
 	if err != nil {
 		return err
@@ -605,6 +621,12 @@ func (p *Probe) wait() {
 func (p *Probe) Start(ctx context.Context, dataChan chan *metrics.EventMetrics) {
 	defer p.wait()
 
+	if p.certinel != nil {
+		go func() {
+			log.Fatal(p.certinel.Start(ctx))
+		}()
+	}
+
 	p.updateTargetsAndStartProbes(ctx, dataChan)
 
 	// Do more frequent listing of targets until we get a non-zero list of

```
That patch is untested, but if the direction looks reasonable I'd be happy to work on a proper PR.

**Comment by manugarg** [comment]
Hi Chris,

Thanks for filing this. Just wanted to let you know that I've seen this and trying to figure out the right approach.

We use a common package right now for TLS config: https://github.com/cloudprober/cloudprober/tree/master/common/tlsconfig. It's used by other features too. I am trying to see how we can incorporate the live-reload functionality in this package itself. Give me a few more days to think about this, I'll get back to you.

Cheers,
Manu

**Comment by cbroglie** [comment]
I was thinking about this some more, perhaps an implementation like this would fit in better:
```diff
diff --git a/common/tlsconfig/tlsconfig.go b/common/tlsconfig/tlsconfig.go
index ec848ef..53d496c 100644
--- a/common/tlsconfig/tlsconfig.go
+++ b/common/tlsconfig/tlsconfig.go
@@ -19,6 +19,8 @@ import (
 	"crypto/tls"
 	"crypto/x509"
 	"fmt"
+	"sync"
+	"time"
 
 	"github.com/cloudprober/cloudprober/common/file"
 	configpb "github.com/cloudprober/cloudprober/common/tlsconfig/proto"
@@ -44,20 +46,44 @@ func UpdateTLSConfig(tlsConfig *tls.Config, c *configpb.TLSConfig) error {
 	}
 
 	if c.GetTlsCertFile() != "" {
-		certPEMBlock, err := file.ReadFile(c.GetTlsCertFile())
-		if err != nil {
-			return fmt.Errorf("common/tlsconfig: error reading TLS cert file (%s): %v", c.GetTlsCertFile(), err)
+		loadCert := func() (*tls.Certificate, error) {
+			certPEMBlock, err := file.ReadFile(c.GetTlsCertFile())
+			if err != nil {
+				return nil, fmt.Errorf("common/tlsconfig: error reading TLS cert file (%s): %v", c.GetTlsCertFile(), err)
+			}
+			keyPEMBlock, err := file.ReadFile(c.GetTlsKeyFile())
+			if err != nil {
+				return nil, fmt.Errorf("common/tlsconfig: error reading TLS key file (%s): %v", c.GetTlsKeyFile(), err)
+			}
+
+			cert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+			if err != nil {
+				return nil, fmt.Errorf("common/tlsconfig: error initializing cert from cert key pair: %v", err)
+			}
+			return &cert, nil
 		}
-		keyPEMBlock, err := file.ReadFile(c.GetTlsKeyFile())
+
+		cert, err := loadCert()
 		if err != nil {
-			return fmt.Errorf("common/tlsconfig: error reading TLS key file (%s): %v", c.GetTlsKeyFile(), err)
+			return err
 		}
+		loadTime := time.Now()
+		reloadInterval := 1 * time.Minute // TODO: make configurable in configpb.TLSConfig
 
-		cert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
-		if err != nil {
-			return fmt.Errorf("common/tlsconfig: error initializing cert from cert key pair: %v", err)
+		var lock sync.Mutex
+		tlsConfig.GetClientCertificate = func(cri *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			lock.Lock()
+			defer lock.Unlock()
+
+			if time.Now().After(loadTime.Add(reloadInterval)) {
+				cert, err = loadCert()
+				if err != nil {
+					return nil, err
+				}
+				loadTime = time.Now()
+			}
+			return cert, nil
 		}
-		tlsConfig.Certificates = append(tlsConfig.Certificates, cert)
 	}
 
 	if c.GetServerName() != "" {

```

**Comment by manugarg** [comment]
Hello @cbroglie,

Sorry for the super delay on this one. I was out on vacation for some time and then got busy with the backlog.

After thinking about various solutions and even doing some implementations, I think something simple like the following might make sense:
https://github.com/cloudprober/cloudprober/pull/428

This is inspired by your second suggestion but just a little bit more generic. Please let me know what do you think and if it makes sense.

**Comment by manugarg** [comment]
@cbroglie I've gone ahead and merged the PR #428, as it seems safe. Whenever you get a chance, please give the latest code a try (either :main tag, or tip release).

**Comment by cbroglie** [comment]
Thanks! And sorry for the delay myself - I left some comments on #428, but it looks like a good approach to me.
